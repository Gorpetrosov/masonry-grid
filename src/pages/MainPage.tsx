import Layout from "../components/layout/Layout.tsx";
import { Navbar } from "../components/navbar/Navbar.tsx";
import {useDispatch, useSelector} from "react-redux";
import {AppDispatch, RootState} from "../store/store.ts";
import {useCallback, useEffect, useMemo, useState} from "react";
import debounce from "lodash.debounce";
import {getMorePhotos, searchPhotos} from "../store/photoSlice.ts";
import {Masonry} from "../components/masonry/Masonry.tsx";

const MainPage = () => {
    const dispatch = useDispatch<AppDispatch>();
    const { loading, entities, total_results, page } = useSelector((state: RootState) => state.photos);
    const [currentQuery, setCurrentQuery] = useState("");
    const [isSearching, setIsSearching] = useState(false);
    const [hasMore, setHasMore] = useState(true);

    useEffect(() => {
        if(!entities.length){
            dispatch(getMorePhotos({ page }));
        }
    }, []);

    useEffect(() => {
        setHasMore(total_results - entities.length > 0);
    }, [total_results, entities.length, setHasMore]);

    const debouncedSearch = useMemo(
        () => debounce(async (query: string) => {
            try {
                await dispatch(searchPhotos({
                    query,
                    page,
                }));
            } finally {
                setIsSearching(false);
            }
        }, 300),
        [dispatch, page]
    );

    const getNewImages = useCallback(() => {

        if(currentQuery){
            dispatch(getMorePhotos({query: currentQuery,page: page + 1 }));
        }else {
            dispatch(getMorePhotos({ page: page + 1 }));
        }
    }, [currentQuery, dispatch, page])

    const handleImagesSearch = useCallback((query: string) => {
        setCurrentQuery(query);
        if (query.trim()) {
            setIsSearching(true);
            debouncedSearch(query.trim());
        }
    }, [debouncedSearch]);

    const updateInputValue = useCallback((query: string) => {
       setCurrentQuery(query.trim());
    },[])

    // Cleanup debounce on unmount
    useEffect(() => {
        return () => {
            debouncedSearch.cancel();
            setIsSearching(false);
        };
    }, [debouncedSearch]);

    return (
        <Layout>
                <Navbar
                    onSearch={handleImagesSearch}
                    isDisabled={loading || isSearching}
                    onInputUpdate={updateInputValue}
                />
                {!loading && !isSearching && entities && (
                   <Masonry photos={entities} loadMore={getNewImages} isLoading={loading} hasMore={hasMore} />
                )}
        </Layout>
    );
};

export default MainPage;