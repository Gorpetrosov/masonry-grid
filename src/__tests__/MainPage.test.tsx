import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import { BrowserRouter } from 'react-router-dom'
import MainPage from '../pages/MainPage.tsx'
import photosReducer from '../store/photoSlice.ts'
import { vi } from 'vitest';

vi.mock('../components/masonry/Masonry.tsx', () => ({
    Masonry: () => <div data-testid="masonry-mock" />
}))

vi.mock('../components/navbar/Navbar.tsx', () => ({
    Navbar: () => <div data-testid="navbar-mock" />
}))

vi.mock('lodash.debounce', () => ({
    default: vi.fn((fn) => fn)
}))

vi.mock('lodash.debounce', () => ({
    default: vi.fn(() => {
        const mocked = vi.fn() as unknown as ((...args: never[]) => never) & { cancel: () => void };
        mocked.cancel = vi.fn();
        return mocked;
    })
}));

describe('MainPage', () => {
    const mockStore = configureStore({
        reducer: {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-expect-error
            photos: photosReducer
        },
        preloadedState: {
            photos: {
                entities: [
                    { id: 1, src: { original: 'img1.jpg' }, alt: 'Image 1', photographer: 'Photog 1' },
                    { id: 2, src: { original: 'img2.jpg' }, alt: 'Image 2', photographer: 'Photog 2' }
                ],
                loading: false,
                total_results: 2,
                page: 1,
                error: null
            }
        }
    })

    afterEach(() => {
        vi.clearAllMocks()
    })

    it('should match snapshot', () => {
        const { container } = render(
            <Provider store={mockStore}>
                <BrowserRouter>
                    <MainPage />
                </BrowserRouter>
            </Provider>
        )

        expect(container).toMatchSnapshot()
    })
})