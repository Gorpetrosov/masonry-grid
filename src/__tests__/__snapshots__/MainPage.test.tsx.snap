// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainPage > should match snapshot 1`] = `
<div>
  <div
    class="layout"
  >
    <header
      class="layout-header"
    >
      <h1>
        Welcome Masonry Grid
      </h1>
    </header>
    <div
      class="layout-main"
    >
      <div
        data-testid="navbar-mock"
      />
      <div
        data-testid="masonry-mock"
      />
    </div>
    <footer
      class="layout-footer"
    >
      <p>
        Â© 
        2025
         
        Masonry APP
      </p>
    </footer>
  </div>
</div>
`;
